### DISCUSSION TITLE
dynamically imported asset's published chunks path in imported compiled script

### DISCUSSION THREAD

**Discussion by shashank-brightness** [webpack/webpack#18318] [Q&A]
Hi, i am trying to import some html from template using dynamic import statement and hence webpack generates its chunk file in the output directory.

Let's say i am importing main.html as a promise and then pass this as parameter to some other function.

```
const loadMainHTML = () => import('./main.html');
import initHTML from '../../initHTML';
initHTML('main-component',loadMainHTML)
```

and for initHTML.js, this is how this promise is being used as `loadComponentHTML` which is this same function `loadMainHTML` which was passed

```
loadComponentHTML().then(function(componentHTML) {
    const newComponent = new DOMParser().parseFromString(
              componentHTML.default,
              'text/html'
            ).body.firstChild;
})
```

and everything is working nicely and webpack published this chunk to separate out html from main file as 
`src_components_main_main_html.js` in the output directory. In my case public directory path is output instead of dist, so chunk is going there

But the thing is that when i try to use compiled output of mypackage to consumer side (where compiled script is imported), then it is not finding the chunk url path correctly.

I am using laravel mix plugin in the project and i need to assign `publicPath` for that one script to output directory inside my package. 

How to set public path to the path of node_modules package file? so let's say if i import `public-path.js` before anything else i would like to set public path to one level up where output directory currently is.

something like `__webpack_public_path__ = '../'; `
so when i import compiled file in consumer end, it points to output dir of package.

**Comment by alexander-akait** [comment]
Why you don't use `publicPath: 'auto'`?

**Reply by shashank-brightness** [reply]
Setting public path to auto results in something even weird path starting from page location and then appending auto to it. 
`'http://localhost:5600/auto./src_components_main_main_html.js' `

I have webpack on consumer side as well but i am compiling script from package itself and exporting compiled file from package.

**Reply by shashank-brightness** [reply]
I need full path to node_modules where my package is but that should be relative to anywhere whether i am using script in my package or import compiled file at consumer.
This is what i eventually need if using at consumer. 
`http://localhost:5600/node_modules/my-package/output/src_components_main_main_html.js`

and if using on mypackage then only
`output/src_components_main_main_html.js`

Ideally want to resolve using same module resolution algorithm which npm uses by default 
e.g. we can import like this
`import 'package/output/mychunkfile';`

but i want to get path of that chunk and not the chunk(module) itself.

**Reply by alexander-akait** [reply]
>I need full path to node_modules where my package is but that should be relative to anywhere whether i am using script in my package or import compiled file at consumer.

For this we have `publicPath: 'auto'`

> 'http://localhost:5600/auto./src_components_main_main_html.js' 

Sounds like you have a bug somewhere

Please create minimum reproducible test repo using github

**Reply by shashank-brightness** [reply]
@alexander-akait 
I don't have bug anywhere but I need to access chunk from my parent app which i can't do right now due to it requiring the incorrect path. 

If I use chunk from the package then i can use it within the package but on my parent app level, asset publishing directory is different than what is for package and I can not somehow access chunks which are generated from bundled file from `node_modules/mypackage/output/js/my_chunk_generated_via_dynamic_imports.js`. 

I was earlier source file as main entry point of my package (src/index.js) in package.json but now i need to use bundled file which is in `output/index.js` so this creates the issue. I will create reproducible example and let you know.

**Reply by alexander-akait** [reply]
> I will create reproducible example and let you know.

Feelf ree to ping me and I will show how you can try to solve it

**Reply by shashank-brightness** [reply]
I am not familiar with how to reproduce this with some online editor to give you example of why this is not working but i hope you already got idea about the problem.

**Reply by shashank-brightness** [reply]
I have created an example [here](https://codesandbox.io/p/devbox/359sqg?migrateFrom=92xj34&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clv4trc4j0006356i1ftjij7w%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clv4trc4j0002356in1858tl4%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clv4trc4j0004356indn8fz27%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clv4trc4j0005356igsk6ldo9%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clv4trc4j0002356in1858tl4%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv4trc4i0001356iakhvslrm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%257D%255D%252C%2522id%2522%253A%2522clv4trc4j0002356in1858tl4%2522%252C%2522activeTabId%2522%253A%2522clv4trc4i0001356iakhvslrm%2522%257D%252C%2522clv4trc4j0005356igsk6ldo9%2522%253A%257B%2522id%2522%253A%2522clv4trc4j0005356igsk6ldo9%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5600%252C%2522id%2522%253A%2522clv4txyxv00go356ifu5qotuq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clv4txyxv00go356ifu5qotuq%2522%257D%252C%2522clv4trc4j0004356indn8fz27%2522%253A%257B%2522id%2522%253A%2522clv4trc4j0004356indn8fz27%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clv4trc4j0003356ieodj7qj8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clv4trctt000cdif83jxd43v8%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clv4trc4j0003356ieodj7qj8%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) as you can see it does not get path from locally installed test Package and rather displays "whatever" as webpack can not resolve the actual path for `node_modules/test-package/output/images/SampleJPGImage_15mbmb.jpg` because that package uses its bundled file as main entry point instead of source file in src folder. It tries to use `https://359sqg-5600.csb.app/output/js/../images/SampleJPGImage_15mbmb.jpg` url as per its own `publicPath` but that image is not showing because that is not available on consumer.

If you go in the `tempPackage` directory whose name in eco-system is `test-package` and run npm start there then it displays image correctly on its own server but this is not something where it is needed. 

So inshort can't access any of `node_modules/test-package/output` directory assets like chunk, fonts etc... and not just image.

Hope it clears now!!
